WireAPI/Sources/WireAPI/APIs/UpdateEventsAPI/Event decoding/UpdateEvent+FeatureConfig.swift

```diff
@@ -24,10 +24,309 @@ extension UpdateEvent {
+            verificationExpiration: payload.config.verificationExpiration
+        )
+    }
+
+    private struct EndToEndIdentityConfig: Decodable {
+
+        let acmeDiscoveryUrl: String?
+        let verificationExpiration: UInt
+
+    }
+
+}
+
+// MARK: - File sharing update
+
+private extension KeyedDecodingContainer<FeatureConfigEventCodingKeys> {
+
+    func decodeFileSharingConfig() throws -> FileSharingFeatureConfig {
+        let payload = try decode(
+            FeatureWithoutConfig.self,
+            forKey: .payload
+        )
+
+        return FileSharingFeatureConfig(status: payload.status)
+    }
+
+}
+
+// MARK: - MLS update
+
+private extension KeyedDecodingContainer<FeatureConfigEventCodingKeys> {
+
+    func decodeMLSConfig() throws -> MLSFeatureConfig {
+        let payload = try decode(
+            FeatureWithConfig<MLSConfig>.self,
+            forKey: .payload
+        )
+
+        return MLSFeatureConfig(
+            status: payload.status,
+            protocolToggleUsers: payload.config.protocolToggleUsers,
+            defaultProtocol: payload.config.defaultProtocol,
+            allowedCipherSuites: payload.config.allowedCipherSuites,
+            defaultCipherSuite: payload.config.defaultCipherSuite,
+            supportedProtocols: payload.config.supportedProtocols
+        )
+    }
+
+    private struct MLSConfig: Decodable {
+
+        let protocolToggleUsers: Set<UUID>
+        let defaultProtocol: MessageProtocol
+        let allowedCipherSuites: [MLSCipherSuite]
+        let defaultCipherSuite: MLSCipherSuite
+        let supportedProtocols: Set<MessageProtocol>
+
+    }
+
+}
+
+// MARK: - MLS migration update
+
+private extension KeyedDecodingContainer<FeatureConfigEventCodingKeys> {
+
+    func decodeMLSMigrationConfig() throws -> MLSMigrationFeatureConfig {
+        let payload = try decode(
+            FeatureWithConfig<MLSMigrationConfig>.self,
+            forKey: .payload
+        )
+
+        return MLSMigrationFeatureConfig(
+            status: payload.status,
+            startTime: payload.config.startTime,
+            finaliseRegardlessAfter: payload.config.finaliseRegardlessAfter
+        )
+    }
+
+    private struct MLSMigrationConfig: Decodable {
+
+        let startTime: Date?
+        let finaliseRegardlessAfter: Date?
+
+    }
+
+}
+
+// MARK: - SelfDeletingMessagesn update
+
+private extension KeyedDecodingContainer<FeatureConfigEventCodingKeys> {
+
+    func decodeSelfDeletingMessagesConfig() throws -> SelfDeletingMessagesFeatureConfig {
+        let payload = try decode(
+            FeatureWithConfig<SelfDeletingMessagesConfig>.self,
+            forKey: .payload
+        )
+
+        return SelfDeletingMessagesFeatureConfig(
+            status: payload.status,
+            enforcedTimeoutSeconds: payload.config.enforcedTimeoutSeconds
+        )
+    }
+
+    private struct SelfDeletingMessagesConfig: Decodable {
+
+        let enforcedTimeoutSeconds: UInt
+
+    }
```

Comment:
I would like to take this example of the pattern used here to question if it's the responsibility of the `KeyedDecodingContainer` to know about the final types and map the decoded values to a non-decodable object?
From my understanding it's rather a concrete container that provides access to the decoded storage.

Therefore my suggestion would be to create a new struct (instead of extending the container)  to provide the required functionality of mapping. It also has the benefits of a pure function without any state or modification from outside:
```swift
private struct SelfDeletingMessagesConfigDecoder {
    private struct Config: Decodable {
        let enforcedTimeoutSeconds: UInt
    }

    func decode(
        from container: KeyedDecodingContainer<FeatureConfigEventCodingKeys>
    ) throws -> SelfDeletingMessagesFeatureConfig {
        let payload = try container.decode(
            FeatureWithConfig<Config>.self,
            forKey: .payload
        )

        return SelfDeletingMessagesFeatureConfig(
            status: payload.status,
            enforcedTimeoutSeconds: payload.config.enforcedTimeoutSeconds
        )
    }
}
```

Calling:
- instead of `try container.decodeSelfDeletingMessagesConfig()`
- one can use `try SelfDeletingMessagesConfigDecoder().decode(from: container)`

Reply:
Good point, I'll make the change across all event decoding.

Reply:
done

Reply:
Amazing, thanks!